File: 알고리즘 스터디 문제들/Part 7 - 백트래킹/문제/N-Queen.py
# 파이썬은 시간 초과 방지를 위해 약간 다른 풀이를 요구한다.
# board[i] == i번째 행에 퀸이 존재하는 열이라 정의하자.
# PyPy3로 제출한다.


def nQueen(r):
    global N, board

    if (r == N):
        return 1

    total = 0
    for c in range(1, N + 1):
        flag = True
        for i in range(1, r + 1):
            if (board[i] == c or abs((r + 1) - i) == abs(c - board[i])):
                flag = False
                break

        if (flag):
            board[r + 1] = c
            total += nQueen(r + 1)

    return total


N = int(input())
board = [0 for i in range(N + 1)]
print(nQueen(0))

File: 알고리즘 스터디 문제들/Part 7 - 백트래킹/문제/N-Queen.cpp
#include <iostream>
#include <array>
using namespace std;

int N;
int ans = 0;
array<array<bool, 15>, 15> board = {}; // 체스판에 퀸이 있는 칸을 확인하는 배열

void nQueen(int x, int y)
{
    // 세로 방향, 가로는 판단하지 않음
    for (int i = 0; i < N; i++)
    {
        if (board[i][x])
            return;
    }

    // 대각선 방향
    for (int i = 0; i < N; i++)
    {
        if (x + i < N && y + i < N)
        {
            if (board[y + i][x + i])
                return;
        }
        if (x - i >= 0 && y + i < N)
        {
            if (board[y + i][x - i])
                return;
        }
        if (x + i < N && y - i >= 0)
        {
            if (board[y - i][x + i])
                return;
        }
        if (x - i >= 0 && y - i >= 0)
        {
            if (board[y - i][x - i])
                return;
        }
    }

    if (y + 1 == N)
    {
        ans++;
        return;
    }

    board[y][x] = true;
    for (size_t i = 0; i < N; i++)
        nQueen(i, y + 1);
    board[y][x] = false;
}

int main()
{
    cin >> N;
    for (size_t i = 0; i < N; i++)
        nQueen(i, 0);
    cout << ans;
}